### PR TITLE
fix: apisix restart failed (#10495)

### PR DESCRIPTION
**Pull request by hanqingwu**
### Description

 when apisix is running and after kill, it return to exit  and skip start new instance 
Fixes # (issue)

fix: apisix restart failed (#10495)

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/cli/ops.lua [hunk 1]
```diff
@@ -916,10 +916,15 @@ local function stop(env)
...
 
+function sleep(n)
+  os.execute("sleep " .. tonumber(n))
+end
+
 local function restart(env)
   -- test configuration
   test(env)
   stop(env)
+  sleep(1)
```
Comment (shreemaan-abhishek) [inline apisix/cli/ops.lua:927]: use ngx.sleep() instead?

**Review by shreemaan-abhishek — CHANGES_REQUESTED** [review]

File: apisix/cli/ops.lua (hunk 1 above)
Comment (hanqingwu) [reply apisix/cli/ops.lua:927]: i have try to instead  ngx.sleep  , but i got a error 

/usr/local/openresty//luajit/bin/luajit: /usr/local/apisix/apisix/cli/ops.lua:930: attempt to index global 'ngx' (a nil value)
stack traceback:
	/usr/local/apisix/apisix/cli/ops.lua:930: in function </usr/local/apisix/apisix/cli/ops.lua:925>
	/usr/local/apisix/apisix/cli/ops.lua:980: in function 'execute'
	/usr/local/apisix/apisix/cli/apisix.lua:40: in main chunk
	[C]: at 0x00405290

so what should i do ?

Comment (shreemaan-abhishek) [reply apisix/cli/ops.lua:927]: never mind, it seems ngx module isn't available during this phase.

**Comment by shreemaan-abhishek** [comment]
please fix the CI

File: apisix/cli/ops.lua [hunk 2]
```diff
@@ -916,10 +916,15 @@ local function stop(env)
...
 
+local function sleep(n)
+  execute("sleep " .. tonumber(n))
+end
+
 local function restart(env)
   -- test configuration
   test(env)
   stop(env)
+  sleep(1)
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:927]: I think just sleeping 1 second is not enough, better adding explicit log message
```
--- a/apisix/cli/ops.lua
+++ b/apisix/cli/ops.lua
@@ -805,8 +805,8 @@ local function start(env, ...)
 
         local ok, err, err_no = signal.kill(pid, signone)
         if ok then
-            print("APISIX is running...")
-            return
+            print("the old process is still running, the new one won't be started!")
+            return
         -- no such process
         elseif err_no ~= errno.ESRCH then
             print(err)
```

File: apisix/cli/ops.lua [hunk 3]
```diff
@@ -916,14 +916,37 @@ local function stop(env)
...
 local function restart(env)
   -- test configuration
   test(env)
   stop(env)
+  wait_stop_finish(env,3)
   start(env)
 end
 
 
+
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:949]: remove this blank

File: apisix/cli/ops.lua [hunk 4]
```diff
@@ -916,14 +916,37 @@ local function stop(env)
 end
 
 
+local function sleep(n)
+  execute("sleep " .. tonumber(n))
+end
+
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:922]: Two blanks are needed between functions

File: apisix/cli/ops.lua [hunk 5]
```diff
@@ -916,14 +916,37 @@ local function stop(env)
...
+local function wait_stop_finish(env, n) 
+
+    for i = 1, n do
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:925]: remove this blank

File: apisix/cli/ops.lua [hunk 6]
```diff
@@ -916,14 +916,37 @@ local function stop(env)
...
+        pid = tonumber(pid)
+        if not pid or pid == 0 then
+            print("old apisix process has stopped")
+            return
+        else
+            sleep(1)
+        end
+    end
+ 
+end
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:938]: remove this blank

File: apisix/cli/ops.lua [hunk 7]
```diff
@@ -916,14 +916,37 @@ local function stop(env)
...
+
+local function wait_stop_finish(env, n) 
+
+    for i = 1, n do
+         -- check running
+        local pid_path = env.apisix_home .. "/logs/nginx.pid"
+        local pid = util.read_file(pid_path)
+        pid = tonumber(pid)
+        if not pid or pid == 0 then
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:930]: when will the pid be 0?

File: apisix/cli/ops.lua [hunk 8]
```diff
@@ -916,14 +916,37 @@ local function stop(env)
...
+        end
+    end
+ 
+end
+
 local function restart(env)
   -- test configuration
   test(env)
   stop(env)
+  wait_stop_finish(env,3)
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:944]:
```suggestion
  wait_old_process_exit(env, 3)
```

**Comment by hanqingwu** [comment]
@shreemaan-abhishek  , can you review it  .

File: apisix/cli/ops.lua [hunk 9]
```diff
@@ -806,6 +806,7 @@ local function start(env, ...)
         local ok, err, err_no = signal.kill(pid, signone)
         if ok then
             print("APISIX is running...")
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:808]: delete this log

**Comment by monkeyDluffy6017** [comment]
please merge the master branch

**Comment by hanqingwu** [comment]
@monkeyDluffy6017 , can you tell me what shoud i do ?      1 change requested by  @shreemaan-abhishek   ?

**Review by shreemaan-abhishek — APPROVED** [review]

**Comment by hanqingwu** [comment]
Hi @monkeyDluffy6017 ,   can you approve reviews ?

File: apisix/cli/ops.lua (hunk 9 above)
Comment (monkeyDluffy6017) [reply apisix/cli/ops.lua:808]: No need to print two identical logs, you can change the two places you metioned

File: apisix/cli/ops.lua (hunk 2 above)
Comment (monkeyDluffy6017) [reply apisix/cli/ops.lua]: you can search the code `APISIX is running`, and replace them all with your new log

File: apisix/cli/ops.lua (hunk 9 above)
Comment (hanqingwu) [reply apisix/cli/ops.lua]: ok, i fix it

File: apisix/cli/ops.lua (hunk 2 above)
Comment (hanqingwu) [reply apisix/cli/ops.lua]: ok , i fix it

File: apisix/cli/ops.lua [hunk 10]
```diff
@@ -916,10 +916,32 @@ local function stop(env)
...
+
+
+local function wait_stop_finish(env, n)
+    for i = 1, n do
+         -- check running
+        local pid_path = env.apisix_home .. "/logs/nginx.pid"
+        local pid = util.read_file(pid_path)
+        pid = tonumber(pid)
+        if not pid then
+            print("old apisix process has stopped")
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:931]: I think there is no need to print this log

File: apisix/cli/ops.lua [hunk 11]
```diff
@@ -916,10 +916,32 @@ local function stop(env)
...
+        end
+    end
+end
+
+
 local function restart(env)
   -- test configuration
   test(env)
   stop(env)
+  wait_stop_finish(env, 3)
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:944]: We have checked if the apisix is running in the function `start()`, can you simplify the logic that is repeated in these 2 functions `wait_stop_finish` and `start`

File: apisix/cli/ops.lua [hunk 12]
```diff
@@ -805,7 +805,7 @@ local function start(env, ...)
 
         local ok, err, err_no = signal.kill(pid, signone)
         if ok then
-            print("APISIX is running...")
+            print("the old process is still running, the new one won't be started!")
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:808]:
```suggestion
            print("the old APISIX is still running, the new one will not start")
```

File: apisix/cli/ops.lua (hunk 11 above)
Comment (hanqingwu) [reply apisix/cli/ops.lua:950]: ok

File: apisix/cli/ops.lua (hunk 12 above)
Comment (hanqingwu) [reply apisix/cli/ops.lua]: ok

File: apisix/cli/ops.lua [hunk 13]
```diff
@@ -792,10 +803,8 @@ local function start(env, ...)
     end
 
     -- check running
-    local pid_path = env.apisix_home .. "/logs/nginx.pid"
-    local pid = util.read_file(pid_path)
-    pid = tonumber(pid)
-    if pid then
+    local running, pid = check_running(env)
```
Comment (monkeyDluffy6017) [inline apisix/cli/ops.lua:806]: I think you don't need to `check_running` two times

Comment (hanqingwu) [reply apisix/cli/ops.lua:806]: i don't know what you mean check_running 2 times,  this is check running and then do something

Comment (monkeyDluffy6017) [reply apisix/cli/ops.lua:806]: you check_running in `wait_stop_finish` and `start`, we can simplify this,  just call `check_running` in `start`, and wait in `check_running`

Comment (hanqingwu) [reply apisix/cli/ops.lua:806]: but here are different logic, `start` check_running if get  true , will return error and do  not wait .

Comment (hanqingwu) [reply apisix/cli/ops.lua:806]: t/cli/test_cmd.sh
```
# check running when run repeatedly
out=$(make run; make run || true)
if ! echo "$out" | grep "the old APISIX is still running"; then
    echo "failed: should find APISIX running"
    exit 1
fi


```

Comment (monkeyDluffy6017) [reply apisix/cli/ops.lua:806]: we can wait in the `check_running`, return true if the old process exits successfully, return false if reach the deadline

Comment (hanqingwu) [reply apisix/cli/ops.lua]: ok , i get you point , i fixed  it .

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]

**Pull request merged**